sql/src/main/java/io/crate/protocols/postgres/types/PGArray.java

```diff
@@ -68,49 +71,125 @@ public Object readBinaryValue(ChannelBuffer buffer, int valueLength) {
+        return values.toArray();
+    }
+
+    private int decodeUTF8Text(byte[] bytes, int startIdx, int endIdx, List<Object> objects) {
+        int valIdx = startIdx;
+        boolean jsonObject = false;
+        for (int i = startIdx; i <= endIdx; i++) {
             byte aByte = bytes[i];
-            if (aByte == '{') {
-                valueStartIdx = i;
-            } else if (aByte == ',' || aByte == '}') {
-                if (bytes[valueStartIdx + 1] == '"') {
-                    values.add(innerType.decodeUTF8Text(Arrays.copyOfRange(bytes, valueStartIdx + 2, i - 1)));
-                } else {
-                    values.add(null);
-                }
-                valueStartIdx = i;
+            switch (aByte) {
+                case '{':
+                    if (i == 0 || bytes[i - 1] != '"') {
+                        if (i > 0) {
+                            // n-dimensions array -> call recursively
+                            List<Object> nestedObjects = new ArrayList<>();
+                            i = decodeUTF8Text(bytes, i + 1, endIdx, nestedObjects);
+                            valIdx = i;
+                            objects.add(nestedObjects.toArray());
+                        } else {
+                            // 1-dimension array -> call recursively
+                            i = decodeUTF8Text(bytes, i + 1, endIdx, objects);
+                        }
+                    } else {
+                        // Start of JSON object
+                        valIdx = i - 1;
+                        jsonObject = true;
+                    }
+                    break;
+                case ',':
+                    if (!jsonObject) {
+                        addObject(bytes, valIdx, i - 1, objects);
+                        valIdx = i + 1;
+                    }
+                    break;
+                case '}':
+                    if (i == endIdx || bytes[i + 1] != '"') {
+                        addObject(bytes, valIdx, i - 1, objects);
+                        return i;
+                    } else {
+                        //End of JSON object
+                        addObject(bytes, valIdx, i + 1, objects);
+                        jsonObject = false;
+                        i ++;
+                        valIdx = i;
+                        if (bytes[i] == '}') { // end of array
+                            return i + 2;
+                        }
+                    }
+            }
+        }
+        return endIdx;
+    }
+
+    // Decode individual inner object
+    private void addObject(byte[] bytes, int startIdx, int endIdx, List<Object> objects) {
+        if (endIdx > startIdx) {
+            byte firstValueByte = bytes[startIdx];
+            if (firstValueByte == '"') {
+                byte[] innerBytes = Arrays.copyOfRange(bytes, startIdx + 1, endIdx);
+                String s = new String(innerBytes, StandardCharsets.UTF_8);
+                s = s.replace("\\\"", "\"");
```

Comment:
I think we should here use custom replacing logic instead, like we do for json arrays for better performance.
At least let's add an todo, and do this asap in a different PR.